pkg/clusteragent/admission/mutate/auto_instrumentation.go

```diff
@@ -189,25 +190,76 @@ func injectApmTelemetryConfig(pod *corev1.Pod) {
 	_ = injectEnv(pod, instrumentationInstallIDEnvVar)
 }
 
-func getAllLibsToInject(registry string) map[string]libInfo {
-	libsToInject := map[string]libInfo{}
+// getLibrariesToInjectForApmInstrumentation returns the list of tracing libraries to inject, when APM Instrumentation is enabled
+// - if apm_config.instrumentation.lib_versions set, returns only tracing libraries from apm_config.instrumentation.lib_versions
+// - if language detection is on and can detect the apps' languages, returns only auto-detected languages
+// - otherwise returns all tracing libraries supported by APM Instrumentation
+func getLibrariesToInjectForApmInstrumentation(pod *corev1.Pod, registry string) ([]libInfo, bool) {
+	autoDetected := false
+
+	// Pinned tracing libraries in APM Instrumentation configuration
+	libsToInject := getPinnedLibraries(registry)
+	if len(libsToInject) > 0 {
+		return libsToInject, autoDetected
+	}
+
+	// Tracing libraries for language detection
+	libsToInject = getLibrariesLanguageDetection(pod, registry)
+	if len(libsToInject) > 0 {
+		autoDetected = true
+		return libsToInject, autoDetected
+	}
+
+	// Latest tracing libraries for all supported languages (java, js, dotnet, python, ruby)
+	libsToInject = getAllLatestLibraries(registry)
+
+	return libsToInject, autoDetected
+}
+
+// getPinnedLibraries returns tracing libraries to inject as configured by apm_config.instrumentation.lib_versions
+func getPinnedLibraries(registry string) []libInfo {
```

Comment:
The output of the method doesn't change for each deployment rt ? if so can we cache this ?

Reply:
You are right, pinned library configuration can be changed only during Cluster Agent installation and is the same for all deployments/pods going through Admission Controller mutating webhook.

I've added a simple thread safe caching mechanism for populating/accessing pinned libraries. Let me know if you think we need something more complicated here.

Reply:
In my other PR https://github.com/DataDog/datadog-agent/pull/22694/ I've created a `APMInstrumentation` struct. We could have a field in the struct with this information and initialize it only once in `NewAPMInstrumentation()`.

We can leave this for another PR after the PR linked above is merged.